sys/kern/src/arch/riscv64.rs

```diff
@@ -0,0 +1,603 @@
+    }
+    fn arg6(&self) -> usize {
+        self.a6 as usize
+    }
+
+    fn syscall_descriptor(&self) -> usize {
+        self.a7 as usize
+    }
+
+    /// Writes syscall return argument 0.
+    fn ret0(&mut self, x: usize) {
+        self.a0 = x as u64
+    }
+    fn ret1(&mut self, x: usize) {
+        self.a1 = x as u64
+    }
+    fn ret2(&mut self, x: usize) {
+        self.a2 = x as u64
+    }
+    fn ret3(&mut self, x: usize) {
+        self.a3 = x as u64
+    }
+    fn ret4(&mut self, x: usize) {
+        self.a4 = x as u64
+    }
+    fn ret5(&mut self, x: usize) {
+        self.a5 = x as u64
+    }
+}
+
+// Because debuggers need to know the clock frequency to set the SWO clock
+// scaler that enables ITM, and because ITM is particularly useful when
+// debugging boot failures, this should be set as early in boot as it can
+// be.
+pub unsafe fn set_clock_freq(tick_divisor: u32) {
+    // TODO switch me to an atomic. Note that this may break Humility.
+    unsafe {
+        CLOCK_FREQ_KHZ = tick_divisor;
+    }
+}
+
+pub fn reinitialize(task: &mut task::Task) {
+    *task.save_mut() = SavedState::default();
+
+    // Set the initial stack pointer, ensuring 16-byte stack alignment as per
+    // the RISC-V callineg convention.
+    let initial_stack: usize = task.descriptor().initial_stack;
+    task.save_mut().sp = initial_stack as u64;
+    uassert!(task.save().sp & 0xF == 0);
+
+    // zap the stack with a distinct pattern
+    for region in task.region_table().iter() {
+        if initial_stack < region.base {
+            continue;
+        }
+        if initial_stack > region.base + region.size {
+            continue;
+        }
+
+        let mut uslice: USlice<usize> = USlice::from_raw(
+            region.base as usize,
+            (initial_stack as usize - region.base as usize) >> 4,
+        )
+        .unwrap_lite();
+
+        let zap = task.try_write(&mut uslice).unwrap_lite();
+        for word in zap.iter_mut() {
+            *word = 0xbaddcafebaddcafe;
+        }
+    }
+    // Set the initial program counter
+    task.save_mut().pc = task.descriptor().entry_point as u64;
+}
+
+pub fn apply_memory_protection(task: &task::Task) {
+    use riscv::register::{Mode, Permission, PmpCfg};
+
+    let null_cfg: PmpCfg = PmpCfg::new(Mode::OFF, Permission::NONE, false);
+
+    for (i, region) in task.region_table().iter().enumerate() {
+        if (region.base == 0x0) && (region.size == 0x20) {
+            continue;
+        }
+        let pmpcfg = {
+            let pmp_perm: Permission = match region.attributes.bits() & 0b111 {
+                0b000 => Permission::NONE,
+                0b001 => Permission::R,
+                0b010 => panic!(),
+                0b011 => Permission::RW,
+                0b100 => Permission::X,
+                0b101 => Permission::RX,
+                0b110 => panic!(),
+                0b111 => Permission::RWX,
+                _ => unreachable!(),
+            };
```

Comment:
Should this be a operation within the riscv crate?  I think converting RWX permissions to a bitmask is pretty standard?
@FawazTirmizi

Reply:
Probably. But out of scope for this change, I think. We can update this function once the riscv crate is updated.